src/app/dim-ui/WeaponGroupingIcon.tsx

```diff
@@ -0,0 +1,104 @@
+import { TagValue } from '@destinyitemmanager/dim-api-types';
+import TagIcon from 'app/inventory/TagIcon';
+import { AmmoIcon } from 'app/item-popup/ItemPopupHeader';
+import { vaultWeaponGroupingSettingSelector } from 'app/settings/vault-grouping';
+import { VaultGroupIconValue } from 'app/shell/item-comparators';
+import { DestinyAmmunitionType } from 'bungie-api-ts/destiny2';
+import { useSelector } from 'react-redux';
+import ElementIcon from './ElementIcon';
+import { getWeaponTypeSvgIconFromCategoryHashes } from './svgs/itemCategory';
+
+const VALID_AMMO_TYPES: Record<DestinyAmmunitionType, boolean> = {
+  [DestinyAmmunitionType.None]: true,
+  [DestinyAmmunitionType.Primary]: true,
+  [DestinyAmmunitionType.Special]: true,
+  [DestinyAmmunitionType.Heavy]: true,
+  [DestinyAmmunitionType.Unknown]: true,
+};
+
+const VALID_TAGS: Record<TagValue, boolean> = {
+  favorite: true,
+  keep: true,
+  infuse: true,
+  archive: true,
+  junk: true,
+};
+
+export default function WeaponGroupingIcon({
+  iconValue,
+  className,
+}: {
+  iconValue: VaultGroupIconValue;
+  className?: string;
+}) {
+  const vaultWeaponGroupingSetting = useSelector(vaultWeaponGroupingSettingSelector);
+
+  if (!vaultWeaponGroupingSetting || typeof iconValue === 'undefined') {
+    return null;
+  }
+
+  if (
+    vaultWeaponGroupingSetting === 'typeName' &&
+    typeof iconValue === 'object' &&
+    // eslint-disable-next-line no-implicit-coercion
+    !!iconValue &&
+    !('hash' in iconValue)
```

Comment:
It can help to write some type predicates for things like this:

```
function isCategoryHashes(iconValue: VaultGroupIconValue): iconValue is number[]) {
  ...
}
```

That said, this is pretty complicated. What if `VaultGroupIconValue` was just a tagged union?:

```
type VaultGroupIconValue = { type: 'itemCategoryHash'; value: number[] } | ...;
```

Then it's super easy to just switch on `iconValue.type` and have all your cases under that.

Reply:
I'd actually considered that approach, so if you're onboard I'll refactor it to work this way. 😊